core/gsb-api/src/api.rs

```diff
@@ -0,0 +1,936 @@
+            }
+        } else {
+            panic!("Expected 404 error");
+        }
+    }
+
+    #[actix_web::test]
+    async fn api_400_error_on_ws_connect_to_incorrectly_encoded_service_address() {
+        let mut api = dummy_api();
+        let (bind_req, service_addr) = bind_get_chunk_service_req(&mut api);
+        let body =
+            verify_bind_service_response(bind_req, vec!["GetChunk".to_string()], &service_addr)
+                .await;
+
+        let services_path = body.listen.unwrap().links.unwrap().messages;
+        let services_path = format!("{services_path}_broken_base64");
+        let ws_frames = api.ws_at(&services_path).await;
+        let expected_err_code = StatusCode::BAD_REQUEST;
+        if let Some(err) = ws_frames.err() {
+            match err {
+                WsClientError::InvalidResponseStatus(code) => {
+                    assert!(matches!(code, _expected_err_code))
+                }
+                other => panic!(
+                    "Expected {:?}, got {:?}",
+                    WsClientError::InvalidResponseStatus(expected_err_code),
+                    other
+                ),
+            }
+        } else {
+            panic!("Expected 404 error");
+        }
+    }
+
+    #[actix_web::test]
+    async fn error_on_post_of_duplicated_service_address() {
+        let mut api = dummy_api();
+        let (bind_req, service_addr) = bind_get_chunk_service_req(&mut api);
+        let _ = verify_bind_service_response(bind_req, vec!["GetChunk".to_string()], &service_addr)
+            .await;
+        let second_bind_response = bind_get_chunk_service_req_w_address(&mut api, service_addr)
+            .0
+            .await
+            .unwrap();
+        assert_eq!(second_bind_response.status(), StatusCode::BAD_REQUEST);
+    }
+
+    #[actix_web::test]
+    async fn ws_close_on_service_delete() {}
+
+    #[actix_web::test]
+    async fn buffering_gsb_msgs_before_ws_connect_test() {
+        let mut api = dummy_api();
+
+        let (bind_req, service_addr) = bind_get_chunk_service_req(&mut api);
+        let body =
+            verify_bind_service_response(bind_req, vec!["GetChunk".to_string()], &service_addr)
+                .await;
+
+        let gsb_endpoint = ya_service_bus::typed::service(&service_addr);
+
+        let services_path = body.listen.unwrap().links.unwrap().messages;
+
+        let (gsb_res, ws_res) = tokio::join!(
+            async {
+                println!("GSB req");
+                let gsb_resp = gsb_endpoint
+                    .call(GetChunk {
+                        offset: u64::MIN,
+                        size: PAYLOAD_LEN as u64,
+                    })
+                    .await;
+                println!("GSB res");
+                gsb_resp
+            },
+            async {
+                println!("WS sleep");
+                tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
These 2 async closures could be done sequentially, I think.
This way we would avoid sleeping arbitrary number of seconds in test, which can turn out to be flaky

Reply:
Did not managed to address it because ya-client task, but it does no harm.